### DISCUSSION TITLE
Unable to configure TLS

### DISCUSSION THREAD

**Discussion by Whitespirit0** [rabbitmq/hop#279] [General]
Hello,

I'm trying to set up TLS with Hop 4.1.0, but I'm unable to import `JdkHttpClientHttpLayer` class in my code.
I guess it's because it's package-private?

2nd question: do the default timeouts defined in `JdkHttpClientHttpLayer` still apply if I follow instructions given in https://github.com/rabbitmq/hop#tls ?

Finally, do the login/password need to be passed twice (once in `.setHeader("Authorization", authorization("guest", "guest")))` and once in `.username("guest").password("guest")`), as described in https://github.com/rabbitmq/hop#advanced-configuration?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
If `JdkHttpClientHttpLayer` is used in examples and this is by design, perhaps it should be made public. @acogoluegnes is it meant to be public?

I cannot find any tests that explicitly use `setHeader` that way, e.g. [see this method](https://github.com/rabbitmq/hop/blob/main/src/test/groovy/com/rabbitmq/http/client/ClientSpec.groovy#L87).

**Comment by acogoluegnes** [comment]
`JdkHttpClientHttpLayer` should be public, this is a leftover from a previous design. I'll push a fix and cut a patch release.

**Comment by acogoluegnes** [comment]
> Finally, do the login/password need to be passed twice (once in `.setHeader("Authorization", authorization("guest", "guest")))` and once in `.username("guest").password("guest")`), as described in https://github.com/rabbitmq/hop#advanced-configuration?

Only if you provide your own `Consumer<HttpRequest.Builder>`. The default one set up the header, so you must do it if you override it.

**Reply by Whitespirit0** [reply]
Ok, thanks for your quick answer.

**Comment by acogoluegnes** [comment]
[4.1.1 is out](https://groups.google.com/g/rabbitmq-users/c/wH0HM1k_GNU/m/mIbVD6DfBAAJ) with the fix.

**Reply by Whitespirit0** [reply]
Great, thanks!

**Comment by Whitespirit0** [comment]
Looking at the code, it seems that the answer to my 2nd question (about timeouts) is:
- Default connect timeout is always applied (not sure about this)
- Default request timeout is applied only if I do not provide my own `Consumer<HttpRequest.Builder>` which is the case in https://github.com/rabbitmq/hop#tls
Am I correct?

**Reply by acogoluegnes** [reply]
The source code for the defaults: https://github.com/rabbitmq/hop/blob/f127807cf7433d9c2996cada16a7441f9e5eb36c/src/main/java/com/rabbitmq/http/client/JdkHttpClientHttpLayer.java#L244-L263

The connect timeout is indeed always set, but can be overridden.

The request timeout and authentication are not set as soon as you provide a `Consumer<HttpRequest.Builder>`.

So when setting up TLS with a `Consumer<Builder>` you do not override any defaults.

**Reply by Whitespirit0** [reply]
Got it, thanks.

**Comment by Whitespirit0** [comment]
@acogoluegnes, it seems that there was another leftover: the `JdkHttpClientHttpLayer.Configuration#create` method is also package-private. Unfortunately, this error was hidden on my side by the previous one.

**Reply by acogoluegnes** [reply]
This is what we call encapsulation  :-)  [4.1.2](https://github.com/rabbitmq/hop/releases/tag/v4.1.2) is out with the fix. Sorry for the trouble.

**Reply by Whitespirit0** [reply]
Ha ha yes. No worries, thanks for your reactivity!
